### PR TITLE
[sdk/python] Remove merge options promotion

### LINKED ISSUE #5639: Bug in merge method for `ResourceOptions`
Given a `ComponentResource` that creates an `AWS` resource and a `Kubernetes` resource.

Given also that the `ComponentResource` is passed a `ResourceOptions` with a keyed `providers` property set for e.g. just `Kubernetes`.

If that `ResourceOptions` object is merged with another `ResourceOptions` object using the `merge` function the `keyed` providers property is collapsed into an unkeyed `provider` (singular) property.

Because this new `provider` property is no longer keyed to just `Kubernetes` the provider would also be used for the AWS resource.

This will result in a fatal error, since the AWS resource does not understand a `Kubernetes` provider.

This directly seems to be done on purpose as can be seen here:

https://github.com/pulumi/pulumi/blob/3d2e31289aba2065853d06024a3da9adfc013202/sdk/python/lib/pulumi/resource.py#L522

### PR DESCRIPTION
**Pull request by iwahbe**
# Description

This is the simple alternative to https://github.com/pulumi/pulumi/pull/8759. I think it is more likely to break existing code, but results in ideal merge behavior. 
The semantics derive from https://github.com/pulumi/pulumi/issues/8796

Fixes #5639

## Checklist

- [X] I have added tests that prove my fix is effective or that my feature works

- [ ] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: sdk/python/lib/test/langhost/component_provider_resolution/__main__.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        }, opts=opts)
+
+class OtherProvider(ProviderResource):
+    def __init__(self, name):
+        super().__init__("other", name)
+
+class CombinedComponent(ComponentResource):
+    def __init__(self, name, opts=None):
+        super().__init__("combined:index:CombinedResource", name, opts=opts)
+        parent = ResourceOptions(parent=self).merge(opts)
```
Comment (t0yv0) [inline sdk/python/lib/test/langhost/component_provider_resolution/__main__.py:38]: Is merge typically called by end-users or we call it in the SDK? checking for refs.

**Comment by t0yv0** [comment]
I found this also:

```
def convert_providers(
        provider: Optional['ProviderResource'],
        providers: Optional[Union[Mapping[str, 'ProviderResource'],
                                  Sequence['ProviderResource']]]) -> Mapping[str, 'ProviderResource']:
    """
    Merge all providers opts (opts.provider and both list and dict forms of opts.providers) into a single dict.
    """
    if provider is not None:
        return convert_providers(None, [provider])

    if providers is None:
        return {}

    if isinstance(providers, Mapping):
        return providers

    result = {}
    for p in providers:
        result[p.package] = p

    return result
```

Does that still look right? AFAIK this only affects remote=True (Multi-Language-Component resources) during processing of register_resource.

File: sdk/python/lib/test/langhost/component_provider_resolution/__main__.py (hunk 1 above)
Comment (iwahbe) [reply sdk/python/lib/test/langhost/component_provider_resolution/__main__.py:38]: We use it in generated SDKs, but I expect end users to call it in the wild. Writing component resources which behave as expected generally involves calling `merge` manually.

**Comment by t0yv0** [comment]
The changes look sane to me given our discussion but I'm a bit worried about cross-language consistency and how this compares to `export function mergeOptions` in Node for example, where `normalizeProviders` promotes 1 providers to provider.

**Comment by t0yv0** [comment]
@iwahbe just to clarify this is ready for final review or you'll be making changes post the design doc discussion? Thanks a lot for taking care of this.

**Comment by iwahbe** [comment]
@t0yv0 I have implemented the changes discussed. This is ready for final review.

File: sdk/python/lib/pulumi/resource.py [hunk 2]
```diff
@@ -796,26 +795,23 @@ def __init__(self,
...
+        if len(type_components) == 3:
+            [pkg, _, _] = type_components
+
+        opts.provider, providers = self._get_providers(t, pkg, opts)
 
         self._protect = bool(opts.protect)
         self._provider = opts.provider if custom else None
+        if self._provider and self._provider.package != pkg:
+            action = (
+                (opts.urn is not None and "get") or
```
Comment (t0yv0) [inline sdk/python/lib/pulumi/resource.py:809]: Oof.

`"get" if opts_run is None else "read" if opts_id is None else "register"`?

File: sdk/python/lib/pulumi/resource.py [hunk 3]
```diff
@@ -832,6 +828,57 @@ def __init__(self,
...
+
+        # The provider supplied by the parent
+
+        # Cast is safe because the `and` will resolve to only None or the result
+        # of get_provider (which is Optional[ProviderResource]).
+        # See https://github.com/python/mypy/issues/12030
+        parent_provider = cast(Optional[ProviderResource], opts.parent and opts.parent.get_provider(t))
+        provider = opts.provider or ambient_provider or parent_provider
+
+        if pkg and opts.provider:
```
Comment (t0yv0) [inline sdk/python/lib/pulumi/resource.py:867]: Condition looks correct.

File: CHANGELOG_PENDING.md [hunk 4]
```diff
@@ -11,6 +11,11 @@
 
 ### Bug Fixes
 
-- [codegen/nodejs] - Generate an install script that runs `pulumi plugin install` with 
+- [sdk/python] - Prevent `ResourceOptions.merge` from promoting between the
+  `.provider` and `.providers` fields. This is a subtle breaking change. This changes
```
Comment (t0yv0) [inline CHANGELOG_PENDING.md:15]: These are user facing docs. I'm wondering if we can't describe coherently a scenario that's breaking and what the user should do should we just omit the words "subtle breaking change"? Can continue presenting it as a bugfix for the interaction of the two fields (that is treat current behavior prior to fix as an error).

**Review by t0yv0 — APPROVED** [review]
LGTM with minor nits.

**Pull request merged**